### PR TITLE
Refactor Heartbeat/Icmp to reduce class dependencies

### PR DESCRIPTION
**Pull request by sancar**
Connection Manager has no dependency to HeartBeatManager
and ClientImcpManager anymore. And they don't depend on
ConnectionManager as well.

### CODE REVIEW

**Review by zpqrtbnk — APPROVED** [review]
Can't find anything wrong, and it certainly looks cleaner.

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/HeartbeatManager.java [hunk 1]
```diff
@@ -16,92 +16,56 @@
...
+                             ILogger logger,
+                             long heartbeatIntervalMillis,
+                             long heartbeatTimeoutMillis,
+                             ActiveConnections activeConnections) {
+        taskScheduler.scheduleWithRepetition(() -> {
+            for (Connection connection : activeConnections.get()) {
+                if (!connection.isAlive()) {
+                    return;
+                }
+                long now = Clock.currentTimeMillis();
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/HeartbeatManager.java:53]: Minor: I don't have some arguments to back that, but I personally liked the version that we get the current time once better

**Review by mdumandag — COMMENTED** [review]
LGTM, I just left some minor comments

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/ActiveConnections.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+package com.hazelcast.client.impl.connection.tcp;
+
+import com.hazelcast.internal.nio.Connection;
+
+import java.util.Collection;
+
+/**
+ * Provides access to current open and authenticated connections to the cluster from the client
+ */
+public interface ActiveConnections {
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/ActiveConnections.java:25]: With an explicit interface, I find it much easier to track where the connections are supplied from. You just need to check where ActiveConnections is implemented. With a `Supplier` you need to walk through the function calls until you find where the Supplier is passed. 

After sleeping on it, I think a Supplier/ActiveConnections even is not needed. 
I am sending another commit which I will remove even `ActiveConnections` and pass the map directly.

File: hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java [hunk 3]
```diff
@@ -421,6 +430,21 @@ public void start() {
         }
     }
 
+    private void startHeartbeat() {
+        long heartbeatTimeout = properties.getPositiveMillisOrDefault(HEARTBEAT_TIMEOUT);
+        long heartbeatInterval = properties.getPositiveMillisOrDefault(HEARTBEAT_INTERVAL);
+        ILogger logger = loggingService.getLogger(HeartbeatManager.class);
+        HeartbeatManager.start(this, executionService, logger,
+                heartbeatInterval, heartbeatTimeout,
+                Collections.unmodifiableCollection(connectionManager.getActiveConnections()));
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java:439]: What happens when a new member joins? I think we have to fetch the active connections, each time we are about to ping

Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java:439]: getActiveConnections does not return copy. The modifications on the map are visible in the returned Collections. So there is no need for asking again to the connection manager.

I see that this is already confusing and may be a good sign to revert the second commit.

**Review by mdumandag — APPROVED** [review]

**Pull request merged**